src/pokemon.c

```diff
@@ -0,0 +1,5567 @@
+
+    return result;
+}
+
+u32 GetMonSpeciesPersonalDataAttribute (int monSpecies, int monPersonalDataAttribute)
+{
+    u32 result;
+    PokemonPersonalData * monPersonalData;
+
+    monPersonalData = GetMonPersonalData(monSpecies, 0);
+    result = GetMonPersonalDataAttribute(monPersonalData, monPersonalDataAttribute);
+
+    FreeMonPersonalData(monPersonalData);
+
+    return result;
+}
+
+u8 GetMonPercentToNextLevel (Pokemon * mon)
+{
+    u16 monSpecies;
+    u8 monLevel;
+    u32 monCurrentLevelExp;
+    u32 monExp;
+    u32 monNextLevelExp;
+    u8 monPercentToNextLevel;
+    BOOL recrypt;
+
+    recrypt = DecryptMon(mon);
+    monSpecies = GetMonData(mon, MON_DATA_SPECIES, NULL);
+    monLevel = GetMonData(mon, MON_DATA_LEVEL, NULL);
+    monCurrentLevelExp = GetMonSpeciesLevelExp(monSpecies, monLevel);
+    monNextLevelExp = GetMonSpeciesLevelExp(monSpecies, monLevel + 1);
+    monExp = GetMonData(mon, MON_DATA_EXP, NULL);
+
+    EncryptMon(mon, recrypt);
+    monPercentToNextLevel = ((monExp - monCurrentLevelExp) * 100) / (monNextLevelExp - monCurrentLevelExp);
+
+    return monPercentToNextLevel;
+}
+
+u32 GetMonExpToNextLevel (Pokemon * mon)
+{
+    return GetBoxMonExpToNextLevel(&mon->box);
+}
+
+u32 GetBoxMonExpToNextLevel (BoxPokemon * boxMon)
+{
+    u16 monSpecies = GetBoxMonData(boxMon, MON_DATA_SPECIES, NULL);
+    u16 monNextlevel = GetBoxMonLevel(boxMon) + 1;
+    u32 monExp = GetBoxMonData(boxMon, MON_DATA_EXP, NULL);
+
+    return GetMonSpeciesLevelExp(monSpecies, monNextlevel) - monExp;
+}
+
+u32 GetMonCurrentLevelExp (Pokemon * mon)
+{
+    return GetMonSpeciesLevelExp(GetMonData(mon, MON_DATA_SPECIES, NULL), GetMonData(mon, MON_DATA_LEVEL, NULL));
+}
+
+u32 GetMonSpeciesLevelExp (int monSpecies, int monLevel)
+{
+    return GetMonExpRateLevelExp(GetMonSpeciesPersonalDataAttribute(monSpecies, MON_DATA_PERSONAL_EXP_RATE), monLevel);
+}
+
+void LoadMonExperienceTable (int monExpRate, u32 * monExpTable)
+{
+    GF_ASSERT(monExpRate < 8);
+    NARC_ReadWholeMemberByIndexPair(monExpTable, NARC_INDEX_POKETOOL__PERSONAL__PL_GROWTBL, monExpRate);
+}
+
+u32 GetMonExpRateLevelExp (int monExpRate, int monLevel)
+{
+    u32 result;
+    u32 * expTable;
+
+    GF_ASSERT(monExpRate < 8);
+    GF_ASSERT(monLevel <= 101);
+
+    expTable = Heap_AllocFromHeap(0, 101 * 4);
+    LoadMonExperienceTable(monExpRate, expTable);
+
+    result = expTable[monLevel];
+    Heap_FreeToHeap(expTable);
+
+    return result;
+}
+
+u32 GetMonLevel (Pokemon * mon)
+{
+    return GetBoxMonLevel(&mon->box);
+}
+
+u32 GetBoxMonLevel (BoxPokemon * boxMon)
+{
+    int monSpecies;
+    u32 monExp;
+    BOOL reencrypt;
+
+    reencrypt = DecryptBoxMon(boxMon);
+    monSpecies = GetBoxMonData(boxMon, MON_DATA_SPECIES, 0);
+    monExp = GetBoxMonData(boxMon, MON_DATA_EXP, 0);
+
+    EncryptBoxMon(boxMon, reencrypt);
+
+    return GetMonSpeciesLevel(monSpecies, monExp);
+}
+
+u32 GetMonSpeciesLevel (u16 monSpecies, u32 monExp)
+{
+    int unused_0;
```

Comment:
You should just be able to get rid of excess variables like these, unless they're needed to match because of a reg-swap.